src/mca/gds/shmem/gds_shmem.c

```diff
@@ -594,34 +595,48 @@ pack_shmem_connection_info(
 }
 
 /**
- * Sets shared-memory connection information from a pmix_buffer_t by
- * unpacking the blob and saving the values for the appropriate rank.
+ * Sets shared-memory connection information from a pmix_kval_t by
+ * unpacking the blob and saving the values for the caller.
  */
 static pmix_status_t
 unpack_shmem_connection_info(
     pmix_gds_shmem_job_t *job,
-    pmix_buffer_t *buffer
+    pmix_kval_t *kvbo
 ) {
     pmix_status_t rc = PMIX_SUCCESS;
 
-    pmix_kval_t kval;
-    PMIX_CONSTRUCT(&kval, pmix_kval_t);
+    // Make sure this is the expected type.
+    if (PMIX_BYTE_OBJECT != kvbo->value->type) {
+        rc = PMIX_ERR_TYPE_MISMATCH;
+        PMIX_ERROR_LOG(rc);
+        return rc;
+    }
 
-    int32_t count = 1;
-    PMIX_BFROPS_UNPACK(
-        rc, pmix_client_globals.myserver,
-        buffer, &kval, &count, PMIX_KVAL
+    pmix_buffer_t buffer;
+    PMIX_CONSTRUCT(&buffer, pmix_buffer_t);
+
+    PMIX_LOAD_BUFFER(
+        pmix_client_globals.myserver,
+        &buffer,
+        kvbo->value->data.bo.bytes,
+        kvbo->value->data.bo.size
     );
 
-    while (PMIX_SUCCESS == rc) {
-        // We only pack string, so make sure this is the right kind of data.
-        if (kval.value->type != PMIX_STRING) {
-            rc = PMIX_ERR_BAD_PARAM;
-            PMIX_ERROR_LOG(rc);
+    pmix_kval_t kv;
+    do {
+        PMIX_CONSTRUCT(&kv, pmix_kval_t);
+
+        int32_t count = 1;
+        PMIX_BFROPS_UNPACK(
+            rc, pmix_client_globals.myserver,
+            &buffer, &kv, &count, PMIX_KVAL
+        );
+        if (PMIX_SUCCESS != rc) {
             break;
         }
-        const char *val = kval.value->data.string;
-        if (pmix_gds_shmem_keys_eq(kval.key, PMIX_GDS_SHMEM_KEY_SEG_PATH)) {
+
+        const char *val = kv.value->data.string;
+        if (pmix_gds_shmem_keys_eq(kv.key, PMIX_GDS_SHMEM_KEY_SEG_PATH)) {
```

Comment:
Just curious: why do we need a special function for checking these keys? Isn't `PMIX_CHECK_KEY(&kv, <key>)` equivalent?

Reply:
It might very well be! Let me take a look. I'm always happy to delete code. Thank you.